src/aapns/connection.py

```diff
@@ -0,0 +1,454 @@
+            if not self.outcome:
+                self.outcome = "Exhausted"
+            raise Closed(self.outcome)
+
+        assert stream_id not in self.channels
+
+        self.last_stream_id_got = stream_id
+
+        channel = self.channels[stream_id] = Channel()
+        self.protocol.send_headers(
+            stream_id, request.header_with(self.host, self.port), end_stream=False
+        )
+        self.protocol.increment_flow_control_window(
+            MAX_RESPONSE_SIZE, stream_id=stream_id
+        )
+        self.protocol.send_data(stream_id, request.body, end_stream=True)
+        self.should_write.set()
+
+        try:
+            while not self.closed:
+                channel.wakeup.clear()
+                with suppress(TimeoutError):
+                    await wait_for(channel.wakeup.wait(), request.deadline - now)
+                now = time()
+                if now > request.deadline:
+                    raise Timeout()
+                for event in channel.events:
+                    if isinstance(event, h2.events.ResponseReceived):
+                        channel.header = dict(event.headers)
+                    elif isinstance(event, h2.events.DataReceived):
+                        channel.body += event.data
+                        if len(channel.body) >= MAX_RESPONSE_SIZE:
+                            raise ResponseTooLarge(f"Larger than {MAX_RESPONSE_SIZE}")
+                    elif isinstance(event, h2.events.StreamEnded):
+                        return Response.new(channel.header, channel.body)
+                    elif isinstance(event, h2.events.StreamReset):
+                        raise StreamReset()
+                del channel.events[:]
+            raise Closed(self.outcome)
+        finally:
+            # FIXME reset the stream, if:
+            # * connection is still alive, and
+            # * the stream didn't end yet
+            del self.channels[stream_id]
+
+    async def close(self):
+        """Terminate the connection and free up the resources"""
+        self.closing = True
+        if not self.outcome:
+            self.outcome = "Closed"
+        try:
+            # FIXME distinguish between cancellation and context exception
+            if self.writer:
+                self.writer.cancel()
+                with suppress(CancelledError):
+                    await self.writer
+
+            if self.reader:
+                self.reader.cancel()
+                with suppress(CancelledError):
+                    await self.reader
+
+            self.closed = True
+
+            # at this point, we must release or cancel all pending requests
+            for channel in self.channels.values():
+                channel.wakeup.set()
+
+            self.write_stream.close()
+            with suppress(SSLError, ConnectionError):
+                await self.write_stream.wait_closed()
+        finally:
+            self.closed = True
+
+    @property
+    def state(self):
```

Comment:
I'd prefer the return type being an enum over a string. helps with typing and makes typos easier to catch statically.

Reply:
This is for human consumption. And at this moment it's unused.